src/nunit.analyzers.tests/DiagnosticSuppressors/DiagnosticsSuppressorAnalyzer.cs

```diff
@@ -0,0 +1,109 @@
+            {
+                Assert.That(suppressor.SupportedSuppressions, Does.Contain(suppressionDescriptor), "Supported Suppression");
+            }
+
+            Compilation compilation = TestHelpers.CreateCompilation(testCode);
+
+            ImmutableArray<Diagnostic> compilationErrors = compilation.GetDiagnostics();
+
+            ImmutableArray<Diagnostic> nonHiddenErrors =
+                compilationErrors.Where(d => d.Severity != DiagnosticSeverity.Hidden)
+                                 .ToImmutableArray();
+
+            ImmutableArray<Diagnostic> suppressibleErrors =
+                nonHiddenErrors.Where(d => suppressor.SupportedSuppressions.Any(s => s.SuppressedDiagnosticId == d.Id))
+                               .ToImmutableArray();
+
+            Assert.That(nonHiddenErrors, Is.EquivalentTo(suppressibleErrors), "Non suppressible errors");
+            Assert.That(suppressibleErrors, Is.Not.Empty, "No errors to suppress");
+
+            var withAnalyzer = compilation.WithAnalyzers(ImmutableArray.Create<DiagnosticAnalyzer>(suppressor));
+
+            ImmutableArray<Diagnostic> analyzerErrors =
+                await withAnalyzer.GetAllDiagnosticsAsync()
+                                  .ConfigureAwait(false);
+
+            nonHiddenErrors = analyzerErrors.Where(d => d.Severity != DiagnosticSeverity.Hidden)
+                                            .ToImmutableArray();
+
+            Assert.That(nonHiddenErrors, Is.Not.Empty);
+
+            Assert.Multiple(() =>
+            {
+                foreach (var error in nonHiddenErrors)
+                {
+                    if (suppressionDescriptor is null)
+                    {
+                        Assert.That(error.IsSuppressed, Is.False, "IsSuppressed: " + error);
+                    }
+                    else
+                    {
+                        Assert.That(error.IsSuppressed, Is.True, "IsSuppressed: " + error);
+                        if (error.IsSuppressed)
+                        {
+                            AssertProgrammaticSuppression(error, suppressionDescriptor);
+                        }
+                    }
+                }
+            });
+        }
+
+        private static void AssertProgrammaticSuppression(Diagnostic error, SuppressionDescriptor suppressionDescriptor)
+        {
+            // We cannot check for the correct suppression nicely.
+            // DiagnosticWithProgrammaticSuppression is a private class.
+            // ProgrammaticSuppressionInfo is an internal class.
+            // Resorting to reflection ...
+            Type errorType = error.GetType();
+            Assert.That(errorType.Name, Is.EqualTo("DiagnosticWithProgrammaticSuppression"));
+
+            var programmaticSuppressionInfoProperty = errorType.GetProperty("ProgrammaticSuppressionInfo",
+                                                                            BindingFlags.Instance | BindingFlags.NonPublic);
```

Comment:
I could not understand why we didn't need to check `programmaticSuppressionInfoProperty` for null, as my VS says that the type is `PropertyInfo?` (which is strange as the VS also states that `GetProperty` returns a `PropertyInfo`), and in the next line it says `'programmaticSuppressionInfoProperty' is not null here.`.

Is this because in ".NET Standard 2.1" the return type of `GetProperty` is `PropertyInfo`, so the compiler assumes that this cannot be null, but since I've also have ".NET 5.0" installed (which has `PropertyInfo?` as the return type), then VS shows me this information?

Reply:
All our target frameworks predate nullability. One way to get the correct annotations in is to use the excellent [TunnelVisionLabs.ReferenceAssemblyAnnotator](https://github.com/tunnelvisionlabs/ReferenceAssemblyAnnotator) to pull in .NET5.0 annotated references.
I did that and guess what, it complained about the `programmaticSuppressionInfoProperty` being null. So good catch by you. It also reported a few more.
I added two commits, one with updating the .csproj files and one with fixing the violations.